---
title: "Analysis Report 1: Comparison of Male and Female Microbial Communities"
author: "Isabella Finney"
date: "October 20, 2017"
output: github_document
bibliography: references.bib
csl: bioinformatics.csl
---

# Introduction

  The skin is the largest organ in the human body. It acts as an effective barrier against pathogen and foreign invasions into the body. Not only is it a defense barrier, but it is also populated by a diverse number of microorganisms that serve different purposes. The skin acts as an ecological system depending on the particular environment the skin is exposed to at different times [@grice2011skin]. The ecological environment that the skin encounters determine the colonization of these microorganisms. Other determining factors are based on things that involve the host, such as gender, age, and hygiene. 
  The development and application of modern molecular methods has made it possible for scientists to more fully understand the diverse communities of bacteria found on all over the body. The use of high-throughput sequencing on 16S ribosomal RNA genes has helped accomplished this goal [@olsen1986microbial]. The significance in the 16s RNA gene is that itâ€™s universally found in all bacteria. This allows scientists to compare the sequences at various levels of taxonomic resolution [@woese1987bacterial]. Through various studies, scientists were able to conclude that different portions of the body have more concentrated portions of certain bacterial microorganisms than other areas on the body. Bacterial community concentrations are often dependent on moist, dry and sebaceous microenvironments. One of the difficulties in determining microorganisms on the skin is due to biases and inefficiency in extraction and sequencing 16S RNA genes. Whole genome shotgun sequencing must also be used in order to determine the full diversity of microorganisms. 
  The areas with the most diversity are drier areas. This portion includes the buttocks, forearm, and portions of the hand. The most common microbiota found in these places are from the phyla Proteobacteria, Actinobacteria, Firmicutes, and Bacteriodetes [@grice2008diversity]. In particular, hands are important to the transfer of microorganisms across surfaces. Through various research, scientists were able to determine that 17% of the phylotypes were the same among the hands of the same individual and 13% was present on hands of different individuals [@fierer2008influence]. The hands also have about less than 150 distinctive bacterial phylotypes present [@stackebrandt1994taxonomic]. Age, gender, exposure to materials, hygiene, and dominant handedness were all contributing factors to the variation of microbiota. When taking a closer look at some of the contributing factors, the differences between males and female hands presented an interesting case. Using the results collected from a study that looked at bacterial communities present on the surface of hands [@fierer2010forensic], I further investigated whether gender differences are a strong factor in the production of certain bacterial communities. Some facts to consider are that statistically, male hands had larger pore size, skin thickness, lower pH, and produced more sebum [@pochi1974endocrinologic]. Taking into account these differences, I hypothesized that females would have a higher presence of diverse microbial communities than men. Investigating different aspects of the dataset may help determine whether gender difference plays a large role in the diversity of bacterial communities. 

# Methods

## Sample origin and sequencing

Add about half a page here. In this section instead of first person (I/we), use Fierer et al., since you'll just be describing what they did, based on the methods in their paper.

## Computational

These are the methods you used. Should probably be at least a half of a page. At a very minimum should include citations for DADA2 [@callahan2016] and phyloseq [@mcmurdie2013]. Note that these don't count towards the five references you need to cite in the introduction.

# Results

```{r load-libraries, message = FALSE}
# Be sure to install these packages before running this script
# They can be installed either with the intall.packages() function
# or with the 'Packages' pane in RStudio

# load general-use packages
library("dplyr")
library("tidyr")
library("knitr")
library("ggplot2")

#load RColorBrewer package
library("RColorBrewer")

# this package allows for the easy inclusion of literature citations in our Rmd
# more info here: https://github.com/crsh/citr
# and here:
# http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html
library("citr")

# These are the primary packages well use to clean and analyze the data
# this package needs to be installed from bioconductor -- it's not on CRAN
# see info here: https://benjjneb.github.io/dada2/dada-installation.html
library("dada2")

# This to export a fasta of our final denoised sequence variants
library("seqinr")

# To install this you have to install from GitHub
# See more info here: https://github.com/leffj/mctoolsr
# run this -- install.packages("devtools")
# and then this -- devtools::install_github("leffj/mctoolsr")
library("mctoolsr")

# And this to visualize our results
# it also needs to be installed from bioconductor
library("phyloseq")
```

```{r extract-sample-and-file-names}
# NOTE: Much of the following follows the DADA2 tutorials available here:
# https://benjjneb.github.io/dada2/tutorial.html
# Accessed October 19, 2017

# set the base path for our input data files
path <- "data/raw_data"

# Sort ensures samples are in order
filenames_forward_reads <- sort(list.files(path, pattern = ".fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq
sample_names <- sapply(strsplit(filenames_forward_reads, "\\."), `[`, 1)

# Specify the full path to each of the filenames_forward_reads
filenames_forward_reads <- file.path(path, filenames_forward_reads)
```

```{r check-quality-plots}
# Plots the quality profiles of all twenty samples
plotQualityProfile(filenames_forward_reads[1:20])
```

We can see from the quality profiles that most reads tend to get pretty bad in quality after around 200 bases. Therefore, we decided to set a maximum acceptable sequence length of 225 bases.

```{r filter-reads}
# Place filtered files in filtered/ subdirectory
# note this will fail if the directory doesn't exist
filter_path <- file.path("output", "filtered")
filtered_reads_path <- file.path(filter_path,
                                 paste0(sample_names,
                                        "_filt.fastq.gz"))

# See ?filterAndTrim for details on the parameters
# See here for adjustments for 454 data:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
filtered_output <- filterAndTrim(fwd = filenames_forward_reads,
                                 filt = filtered_reads_path,
                                 maxLen = 225,
                                 maxN = 0, # discard any seqs with Ns
                                 maxEE = 3, # allow w/ up to 3 expected errors
                                 truncQ = 2, # cut off if quality gets this low
                                 rm.phix = TRUE,
                                 compress = TRUE,
                                 multithread = FALSE)
```

```{r filtered-read-counts-table}
# produce nicely-formatted markdown table of read counts
# before/after trimming
kable(filtered_output,
      col.names = c("Reads In",
                    "Reads Out"))
```

```{r learn-errors}
# this build error models from each of the samples
errors_forward_reads <- learnErrors(filtered_reads_path,
                                    multithread = FALSE)
```

```{r visualize-errors-with-plots}
# quick check to see if error models match data
# (black lines match black points) and are generally decresing left to right
plotErrors(errors_forward_reads,
           nominalQ = TRUE)
```

```{r dereplicate-sequences}
# get rid of any duplicated sequences
dereplicated_forward_reads <- derepFastq(filtered_reads_path,
                                         verbose = TRUE)

# Name the derep-class objects by the sample names
names(dereplicated_forward_reads) <- sample_names
```

```{r run-dada}
# parameters adjusted based on recommendations for 454 data here:
# https://benjjneb.github.io/dada2/
#     faq.html#can-i-use-dada2-with-my-454-or-ion-torrent-data
dada_forward_reads <- dada(dereplicated_forward_reads,
                           err = errors_forward_reads,
                           HOMOPOLYMER_GAP_PENALTY = -1, # reduce penalty bc 454
                           BAND_SIZE = 32) # performs local alignments bc indels

# check dada results
dada_forward_reads
```

```{r make-sequence-table}
# produce the 'site by species matrix'
sequence_table <- makeSequenceTable(dada_forward_reads)
```

The output table has `r nrow(sequence_table)` rows (samples) and `r ncol(sequence_table)` columns (sequence variants). Notice how we can embed R code directly in our markdown text.

```{r histogram-of-sequence-lengths}
# Quick check to look at distribution of trimmed and denoised sequences
hist(nchar(getSequences(sequence_table)),
     main = "Histogram of fungal sequence variant lengths",
     xlab = "Sequence length in bp")
```

```{r remove-chimeras}
# Check for and remove chimeras
sequence_table_nochim <- removeBimeraDenovo(sequence_table,
                                            method = "consensus",
                                            multithread = FALSE,
                                            verbose = TRUE)

# What percent of our reads are non-chimeric?
non_chimeric_reads <- round(sum(sequence_table_nochim) / sum(sequence_table),
                            digits = 4) * 100
```

After removing chimeras, we were left with `r non_chimeric_reads`% of our cleaned reads.

```{r table-of-pipeline-read-counts}
# Build a table showing how many sequences remain at each step of the pipeline
get_n <- function(x) sum(getUniques(x)) # make a quick function
track <- cbind(filtered_output, # already has 2 columns
               sapply(dada_forward_reads, get_n),
               rowSums(sequence_table),
               rowSums(sequence_table_nochim))

# add nice meaningful column names
colnames(track) <- c("Input",
                     "Filtered",
                     "Denoised",
                     "Sequence Table",
                     "Non-chimeric")

# set the proper rownames
rownames(track) <- sample_names

# produce nice markdown table of progress through the pipeline
kable(track)
```

```{r assign-taxonomy}
# assigns taxonomy to each sequence variant based on a supplied training set
# made up of known sequences
taxa <- assignTaxonomy(sequence_table_nochim,
                       "data/training/rdp_train_set_16.fa.gz",
                       multithread = FALSE,
                       tryRC = TRUE) # also check with seq reverse compliments

# show the results of the taxonomy assignment
unname(taxa)
```

```{r extract-sequences-to-fasta}
# we want to export the cleaned, trimmed, filtered, denoised sequence variants
# so that we can build a phylogeny - we'll build the phylogeny outside of R
# but we need the fasta file to do so. We keep the names of each sequence as the
# sequence itself (which is rather confusing), because that's how DADA2 labels
# it's columns (e.g. 'species')
# function taken from https://github.com/benjjneb/dada2/issues/88
export_taxa_table_and_seqs <- function(sequence_table_nochim,
                                       file_seqtab,
                                       file_seqs) {
  seqtab_t <- as.data.frame(t(sequence_table_nochim)) # transpose to data frame
  seqs <- row.names(seqtab_t) # extract rownames
  row.names(seqtab_t) <- seqs # set rownames to sequences
  outlist <- list(data_loaded = seqtab_t)
  mctoolsr::export_taxa_table(outlist, file_seqtab) # write out an OTU table
  seqs <- as.list(seqs)
  seqinr::write.fasta(seqs, row.names(seqtab_t), file_seqs) # write out fasta
}

# actually run the function, with the names of the files we want it to create
# and where to put them
export_taxa_table_and_seqs(sequence_table_nochim,
                           "output/sequence_variants_table.txt",
                           "output/sequence_variants_seqs.fa")
```


```{r read-in-metadata-and-create-phyloseq}
# Next we want to read in the metadata file so we can add that in too
# This is not a csv file, so we have to use a slightly different syntax
# here the `sep = "\t"` tells the function that the data are tab-delimited
# and the `stringsAsFactors = FALSE` tells it not to assume that things are
# categorical variables
metadata_in <- read.table(paste0("data/metadata/",
                                 "fierer_forensic_hand_mouse_SraRunTable.txt"),
                          sep = "\t",
                          header = TRUE,
                          stringsAsFactors = FALSE,
                          row.names = 6) # sets sample IDs to row names

# read in the phylogeny, which was created from the fasta exported above
# in Geneious by aligning the sequences with MAFFT and then building a
# Maximum-Likelihood tree with RAxML
tree_in <- read_tree("output/sequence_variants_MAFFT_RAxML.newick")

# Construct phyloseq object (straightforward from dada2 outputs)
phyloseq_obj <- phyloseq(otu_table(sequence_table_nochim,
                                   taxa_are_rows = FALSE), # sample-spp matrix
                         sample_data(metadata_in), # metadata for each sample
                         tax_table(taxa), # taxonomy for each sequence variant
                         phy_tree(tree_in)) # phylogeny from sequence variants

melted_obj <- psmelt(phyloseq_obj)
```


```{r example-phyloseq-plot-1}
# alpha diversity metrics
plot_richness(phyloseq_obj,
              x = "env_material_s",
              measures = c("Shannon"),
              color = "sex_s") +
  xlab("Sample origin") +
  geom_jitter(width = 0.2) +
  theme_bw()
```

**Figure 1**: Alpha diversity measures of the two sample types, colored by gender.

```{r plot-richness-for-male-and-female}
pruned_phyloseq <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
plot_richness(pruned_phyloseq,
              measures = "Observed",
              color = "sex_s")
```

**Figure 2**: Alpha richness of the observed samples colored by gender.

```{r plot-ordination-of-taxa-by-sex}
set.seed(15)
#Pruning to subset the dataset
pruned_phyloseq <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
pruned_ord <- ordinate(pruned_phyloseq, "NMDS", "bray")

p1 <- plot_ordination(pruned_phyloseq, pruned_ord,
                     type = "samples",
                     color = "sex_s",
                     title = "Ordination of Male versus Female Sequences")
print(p1)
```

**Figure 3**: Plot ordination of the taxa present by gender. 

```{r plot-ordination-of-phylum-by-sex}
#Create Plot Ordination showing the phylum present by sex
pruned_phyloseq_ord <- ordinate(pruned_phyloseq, "NMDS", "bray")
p1 <- plot_ordination(pruned_phyloseq, pruned_phyloseq_ord,
                      type = "Taxa", color = "Phylum", title = "taxa")
print(p1)
```

**Figure 4**: Plot Ordination of the phylum present by gender.

```{r example-phyloseq-plot-2}
# phylogeny, yay!
phyloseq_obj <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
plot_tree(phyloseq_obj,
          color = "sex_s",
          ladderize = TRUE) # this arranges the tree branches from short to long
```

**Figure 5**: Inferred phylogeny of sequences, with points on tips representing samples within which each particular taxa occurred. Tree represents maximum likelihood phylogeny inferred using RAxML.

install.packages("RColorBrewer")

```{r bar-plot-of-phlyum-presence-by-sex}
plot_bar(pruned_phyloseq, x = "sex_s", y = "Abundance", fill = "Phylum",
 title = NULL, facet_grid = NULL) +
  scale_fill_brewer(palette = "Paired")
```

**Figure 6**: Bar plot of the phylum present by gender.

```{r bar-plot-of-class-presence-by-sex}
pruned_phyloseq <- subset_samples(phyloseq_obj, sex_s != "Not applicable")
plot_bar(pruned_phyloseq, x = "Class", fill = "sex_s")
```

**Figure 7**: Bar plot of class present colored by gender.

# Discussion

Add around 2-3 pages interpreting your results and considering future directions one might take in analyzing these data.

# Sources Cited


